### PR TITLE
Tmp compute

### PR DESCRIPTION
**Pull request by lixinqi**
在loss.backward内部将loss所在的stream切换为新定义的StreamRole::kTmpCompute，这样能极大地减少main线程blocking wait的时间。

### CODE REVIEW

**Comment by yuanms2** [comment]
> 在loss.backward内部将loss所在的stream切换为新定义的StreamRole::kTmpCompute，这样能极大地减少main线程blocking wait的时间。

能不能解释的更详细一点。loss 的 output 放到一个单独的stream里，为什么就能减少main 线程的等待时间，是main 线程后续的计算instruction 就不需要等待loss 的计算结果吗？在一个单独的stream里算出loss 后是在后续一个合适的时机交给main 线程吗？

File: oneflow/core/autograd/autograd_engine.cpp [hunk 1]
```diff
@@ -123,7 +151,11 @@ Maybe<void> AutogradEngine::RunBackwardAndSaveGrads4LeafTensorIf(const TensorTup
   JUST(CheckConsistentTensorsMeta(outputs));
   JUST(CheckConsistentTensorsMeta(out_grads));
   DisableCheckConsistentTensorMetaScope disable_meta_check;
-  return RunBackwardAndSaveGrads4LeafTensor(outputs, out_grads, retain_graph, create_graph);
+  // Put outputs into kTmpCompute stream for reducing blocking time of outputs[i].numpy() in main
```
[the remark this answers is not in the document]
Comment (lixinqi) [reply oneflow/core/autograd/autograd_engine.cpp:154]: 是main线程。scheduler总是不会blocking的。

File: oneflow/core/autograd/autograd_engine.cpp [hunk 2]
```diff
@@ -123,7 +151,11 @@ Maybe<void> AutogradEngine::RunBackwardAndSaveGrads4LeafTensorIf(const TensorTup
...
+  // thread.
+  auto copied_outputs = JUST(TryCopyForSmallTensor(outputs));
+  JUST(TouchInTmpComputeStream(outputs));
```
[the remark this answers is not in the document]
Comment (lixinqi) [reply oneflow/core/autograd/autograd_engine.cpp:157]: > 原先 numpy 指令和后向指令在同一个 stream 上，所以 scheduler 要先处理完后向指令，然后才能轮到处理 .numpy() 指令

这是事实。

> 现在把 loss tensor 放在了另一个 stream 上，.numpy() 指令就不用和后向指令在一个队里排队了。但我看 .numpy() 对应的 AccessBlobByCallback 指令总是在 Compute stream 上做的，和 tensor 的 last_used_stream 无关

首先，cuda_tensor.numpy()内部实际发射了三条指令：RecordEvent/OpCall(copy_d2h)/AccessBlobByCallback，他们分别在cuda_compute/cuda_d2h/cpu_compute这三条stream上。把cuda_tensor放到cuda_tmp_compute的好处就是RecordEvent不会排在bw_ops + optm_ops之后，所以能很快执行完成。

**Comment by lixinqi** [comment]
> > 在loss.backward内部将loss所在的stream切换为新定义的StreamRole::kTmpCompute，这样能极大地减少main线程blocking wait的时间。
> 
> 能不能解释的更详细一点。loss 的 output 放到一个单独的stream里，为什么就能减少main 线程的等待时间，是main 线程后续的计算instruction 就不需要等待loss 的计算结果吗？在一个单独的stream里算出loss 后是在后续一个合适的时机交给main 线程吗？

https://github.com/Oneflow-Inc/oneflow/pull/8570#discussion_r914707152

**Comment by lixinqi** [comment]
本pr的效果在这里有展示：
https://github.com/Oneflow-Inc/OneTeam/issues/1552#issuecomment-1189070929

File: oneflow/core/framework/stream_on_independent_thread.h [hunk 3]
```diff
@@ -24,13 +24,14 @@ namespace oneflow {
 struct StreamOnIndependentThread : public StreamRoleVisitor<StreamOnIndependentThread> {
   static bool VisitCompute() { return false; }
   static bool VisitHost2Device() { return false; }
-  static bool VisitDevice2Host() { return false; }
+  static bool VisitDevice2Host() { return true; }
```
Comment (daquexian) [inline oneflow/core/framework/stream_on_independent_thread.h:27]: 这里的修改和本次 PR 有关吗

**Review by daquexian — APPROVED** [review]

File: oneflow/core/framework/stream_on_independent_thread.h (hunk 3 above)
Comment (lixinqi) [reply oneflow/core/framework/stream_on_independent_thread.h:27]: 有关。如果D2H不是在单独的线程上，还是和compute stream挤占一个线程。不会有加速

Comment (lixinqi) [reply oneflow/core/framework/stream_on_independent_thread.h]: 我仔细想了一下，这里确实有点牵强。本次Pr不应该和Device2Host有关。因为此处一旦改成了true，会导致原本的一些代码莫名其妙地变慢。
为此，我们再新提AsyncedDevice2Host，它是独立线程的。它的代码在这个commit里https://github.com/Oneflow-Inc/oneflow/commit/f03fe1cd30e88c0d95b9b939295b60692eef132c 。这个StreamRole::kAsyncedDevice2Host的推导在这里 https://github.com/Oneflow-Inc/oneflow/pull/8570/commits/f03fe1cd30e88c0d95b9b939295b60692eef132c#diff-1da8f6bfd2e1c9c7b5d8ceae1ea3e4f47c3dcaf596ab3c00c793f5d5145cfa39R30

**Review by clackhan — APPROVED** [review]
